Implement `OrderedOptions#dig!`

sinsoku opened the pull request:
### Motivation / Background

We often use `dig` with credentials and custom configurations, but the method
returns `nil` instead of an error if a non-existent key is specified by a typo.

It is possible to write without dig to quickly notice a wrong key, but this is tedious.

```ruby
# Before
Rails.application.credentials.fetch(:aws).fetch(:access_key_id)
Rails.application.config_for(:example).fetch(:foo).fetch(:bar)
```

To make this simple to write, this commit adds a bang version method.

```ruby
# After
Rails.application.credentials.dig!(:aws, :access_key_id)
Rails.application.config_for(:example).dig!(:foo, :bar)
```

### Detail

This Pull Request implements `OrderedOptions#dig!` to raise an exception if a non-existent key is specified.

### Additional information

There was discussion about a similar feature in Ruby, but it was last updated a year ago.

- [Version of dig that raises error if a key is not present](https://bugs.ruby-lang.org/issues/14602)
- [#dig that throws an exception if a key doesn't exist](https://bugs.ruby-lang.org/issues/15563)

In my use case, `OrderedOptions#dig!` is sufficient, so I avoided `Hash#dig!` which has a large impact.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

sinsoku committed:
Implement `OrderedOptions#dig!`

We often use `dig` with credentials and custom configurations, but the method
returns `nil` instead of an error if a non-existent key is specified by a typo.

It is possible to write without dig to quickly notice a wrong key, but this is tedious.

```ruby
# Before
Rails.application.credentials.fetch(:aws).fetch(:access_key_id)
Rails.application.config_for(:example).fetch(:foo).fetch(:bar)
```

To make this simple to write, this commit adds a bang version method.

```ruby
# After
Rails.application.credentials.dig!(:aws, :access_key_id)
Rails.application.config_for(:example).dig!(:foo, :bar)
```

MatheusRich commented:
Does `Rails.application.credentials.aws!.access_key_id!` work?

sinsoku commented:
Thank you, I didn't know it worked with credentials.
Since it does not work in custom configurations, I thought this would not work either.

```ruby
Rails.application.credentials.aws!.access_key_id!
#=> 123

Rails.application.config_for(:example).foo!.bar!
(irb):12:in `<main>': undefined method `bar!' for {:bar=>{:baz=>1, :qux=>2}}:Hash (NoMethodError)

Rails.application.config_for(:example).foo!.bar!
                                           ^^^^^
```

I noticed that the credentials are deep_transform [here](https://github.com/rails/rails/blob/a60a5d0f5bbe30a33f8562d77277ef19906944aa/activesupport/lib/active_support/encrypted_configuration.rb#L84-L92).

Is it better to deep_transform also in `config_for` than adding `dig!` method?

MatheusRich commented:
I'd ask one of the maintainers to be sure, but, **in _my_ opinion**, those things should be consistent. I'd love to see that working for custom configurations too.

adrianna-chang-shopify commented on activesupport/test/ordered_options_test.rb:
```diff
@@ -36,14 +36,24 @@ def test_looping
     assert_equal 56, a["test_key"]
     assert_equal 56, a.dig(:test_key)
     assert_equal 56, a.dig("test_key")
+    assert_equal nil, a.dig(:non_existing_key)
```
```suggestion
    assert_nil a.dig(:non_existing_key)
```

sinsoku replied:
Thanks for a review. I changed to `assert_nil` .

sinsoku committed:
Use assert_nil instead of assert_equal

sinsoku committed:
Modify a behavior for consistency

Following changes in the dig method, nested hashes and arrays are no longer converted to symbols.
refs: 1b211421cd0f6fa241a437e734371dd825f4a7f0
